File: frontend/frontend/formfilling/pages.js
const WebPage = require("./page");
const { print } = require("./libs/output");

class WebPages {
    constructor(pages, index = 0) {
        this.pages = pages;
        this.index = index;
    }

    // The class implements the iterator protocol
    [Symbol.iterator]() {
        return this;
    }

    next() {
        if (this.index >= this.pages.length) {
            return { done: true };
        }
        const value = this.pages[this.index];
        this.index++;
        return { value, done: false };
    }

    toString() {
        return this.pages.map((page) => page.name).join(", ");
    }

    get_web_page_by_name(name) {
        return this.pages.find((page) => page.name === name) || null;
    }

    // Skip forward to the page with the given name, ignore all pages' make_actions() methods
    skip_forward_to(name) {
        for (const webPage of this.pages) {
            if (webPage.name === name) {
                return;
            }
            webPage.next();
        }
    }

    // Skip backward to the page with the given name, ignore all pages' make_actions() methods
    skip_backward_to(name) {
        for (const webPage of this.pages) {
            if (webPage.name === name) {
                return;
            }
            webPage.previous();
        }
    }

    // print out all page index, name and description
    print_page_list() {
        print("\nPage list\n", "info");
        for (const index in this.pages) {
            let row = `${index}: ${this.pages[index].name} - ${this.pages[index].description}`;
            print(row);
        }
    }
}

module.exports = WebPages;

File: client/webform/pages.py
from client.webform.page import WebPage
from dataclasses import dataclass

""" Webpages actually represents a program's web forms filling workflow """


@dataclass
class WebPages:
    pages: list[WebPage]
    index: int = 0

    """ The class implements the iterator protocol"""

    def __iter__(self):
        return self

    def __next__(self):
        if self.index >= len(self.pages):
            raise StopIteration
        value = self.pages[self.index]
        self.index += 1
        return value

    def __str__(self):
        return ", ".join([page.name for page in self.pages])

    def get_web_page_by_name(self, name):
        for web_page in self.pages:
            if web_page.name == name:
                return web_page
        return None

    """ Skip forward to the page with the given name, ignore all pages' make_actions() methods "" """ ""

    def skip_forward_to(self, name):
        for web_page in self.pages:
            if web_page.name == name:
                return
            web_page.next()

    """ Skip backward to the page with the given name, ignore all pages' make_actions() methods "" """ ""

    def skip_backward_to(self, name):
        for web_page in self.pages:
            if web_page.name == name:
                return
            web_page.previous()
